[SofaDenseSolver] Fix compilation with SOFA_FLOATING_POINT_TYPE as float

FabienPean opened the pull request:
issue sofa-framework/sofa#2 Fix compilation when option SOFA_FLOATING_POINT_TYPE in CMake selected as float

The problem arising when trying to compile with float was due to missing declaration from newmat external library.
The library has been modified for sofa to use a #define SOFA_FLOAT in include.h
However the SOFA_FLOAT was never defined for the extlib since the current procedure is to set CMake variable SOFA_FLOAT and SOFA_DOUBLE to 1. This was done anyway AFTER processing the cmake of the extlibs.

Therefore, this pr moves the option for floating point type in sofa in the root CMakeLists and adds the #define for the newmat library to actually create the float type functions.

______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] has been reviewed and agreed to be transitional.
- [x] is more than 1 week old (or has fast-merge label).
- [x] reports important changes in Changelog.

**Reviewers will merge only if all these checks are true.**

FabienPean committed:
[SofaDenseSolver] issue#2 Fix compilation when option SOFA_FLOATING_POINT_TYPE in CMake selected as float

damienmarchal commented:
[ci-build]

guparan commented:
Dashboard details are here: https://www.sofa-framework.org/dash/index.php?branch=pr%2Ffix%2Fissue%232

guparan commented:
Hi @FabienPean, 
Any idea why RigidLinearDeformationMappings_test is failing on Windows ?
See [windows7_VS-2015_amd64_origin_options tests report](https://ci.inria.fr/sofa-ci/job/windows7_VS-2015_amd64_pr_2_options/266/testReport/).

EDIT: This is certainly nothing related to your PR. It is also failing for sofa-framework/sofa#261.

EDIT 2: The problem is known, see issue sofa-framework/Flexible#4.

matthieu-nesme commented on SofaKernel/extlibs/newmat/CMakeLists.txt:
```diff
@@ -42,6 +42,9 @@ add_library(${PROJECT_NAME} STATIC ${HEADER_FILES} ${SOURCE_FILES})
 target_compile_definitions(${PROJECT_NAME} PRIVATE "use_namespace")
+if(SOFA_FLOAT)
+    target_compile_definitions(${PROJECT_NAME} PRIVATE "SOFA_FLOAT")
+endif()
```
I do not understand why this is necessary?
Is a dependency to a sofa component missing?

matthieu-nesme replied:
oh it is an extlib....
using a sofa macro in an extlib????

FabienPean replied:
I initially tried the method advocated on the newmat library FAQ [see 2.3 Customizing](http://robertnz.net/nm11.htm). It was not working, because the newmat/include.h file was modified by sofa developers at some point, who also added the sofa macro inside, see line 31. In the end, SOFA_FLOAT in include.h is only a proxy to other define USING_DOUBLE and USING_FLOAT, so it could probably be removed.

matthieu-nesme replied:
hum, looks like something has to be cleaned more deeply in the sofa cmake files to integrate newmat the right way.

FabienPean commented:
I removed the sofa define from newmat/include.h and corrected the CMakeFiles. Now it activates float only when "float" is explicitly selected in root CMakeFiles

FabienPean commented:
Ok I did not expect this problem. I understand now better what Matthieu meant.

When I removed SOFA_DOUBLE from newmat, the target_compile_definitions are only visible within newmat project with the PRIVATE attribute, which was causing problem when the newmat.h was used included in SofaDenseSolver, not having the compile definitions set. It works if the compile definition of newmat are PUBLIC, but I do not know the side effects it may cause.

FabienPean committed:
[Extlib] Update newmat.h to be closer to the original version.

Original version available here [newmat](http://www.robertnz.net/nm_intro.htm)

FabienPean committed:
[Extlib] Fix CMakeLists.txt of newmat extlib

damienmarchal commented:
Thanks Fabian for the quick update of the PR...
[ci-build]

guparan commented:
Did you try compiling with `SOFA_FLOATING_POINT_TYPE = float`? 
It is still failing on my side (Windows + VS-2015).

The Dashboard is OK but compiling with `SOFA_FLOATING_POINT_TYPE = both`.

FabienPean committed:
Try to fix tests to work with precision "float"

FabienPean committed:
[SofaEigen2Solver] Fix class EigenVector to work with type float

FabienPean committed:
[SofaBaseMechanics] Fix UniformMass/DiagonalMass to work with type float

FabienPean committed:
[SofaGeneralLinearSolver] Fix MinResLinearSolver to work with type float

FabienPean committed:
[Gui] Fix QtViewer to work with type float

FabienPean commented:
I did not manage to compile it fully. SofaDenseSolver is now ok but not the remaining of Sofa. I started to fix some errors, however it seems that this feature was not actively maintained recently. 
With latest commit it should compile but it gives weird runtime visualization. (Windows VS2015 also)
![image](https://cloud.githubusercontent.com/assets/9134757/26287521/acf354d4-3e7d-11e7-8e6f-bc5d1a2e6ca2.png)
![image](https://cloud.githubusercontent.com/assets/9134757/26287526/b7ce280c-3e7d-11e7-8262-c22a7fc16069.png)
![image](https://cloud.githubusercontent.com/assets/9134757/26287527/c4514e1a-3e7d-11e7-96e3-a18bcf007403.png)

If the feature is a must-have, maybe the CI should test with float-only and double-only instead of "both" ?

guparan commented:
Thank you for your fixes @FabienPean !

Two remarks about consistency:
- `Sofa_test` default template is `SReal`. I would go for `Sofa_tests<>` inheritance. Unless there is a specific reason for inheriting from `Sofa_tests<SReal>` ?
- `#ifndef SOFA_FLOAT` block usually comes before `#ifndef SOFA_DOUBLE` block.

I can push these tiny corrections if you're ok 😉

FabienPean commented:
Hi Guillaume, sure feel free ! :)

damienmarchal commented:
The 'leaking' snake looks great :)

guparan committed:
[Tests] inherit from Sofa_test<> instead of Sofa_test<SReal>

guparan committed:
[All] Move "#ifndef SOFA_FLOAT" before "#ifndef SOFA_DOUBLE"

guparan committed:
[Tests] Remove useless using

guparan committed:
[SofaGeneralLinearSolver] ADD float/double separation in MinResLinearSolver

guparan committed:
[SofaKernel] Tabs to spaces in DiagonalMass

guparan commented:
Here they are !

damienmarchal commented:
It is unclear to me if it is ready. Isn't the screenshot from @FabienPean showing something is broken ?

guparan commented:
At least the compilation is fixed so I'd say this PR does its job. I had the same visualization glitches on my side though. Maybe a quick review to ensure we did not create the problem - @fredroy ?

matthieu-nesme commented on SofaKernel/extlibs/newmat/CMakeLists.txt:
```diff
@@ -41,7 +41,12 @@ set(SOURCE_FILES
 target_include_directories(${PROJECT_NAME} PUBLIC "$<INSTALL_INTERFACE:include>")
-target_compile_definitions(${PROJECT_NAME} PRIVATE "use_namespace")
+target_compile_definitions(${PROJECT_NAME} PUBLIC "use_namespace")
+if(NOT DEFINED SOFA_WITH_DOUBLE)
```
To be fully clean, it would be great to add this code where the newmat dir in included in cmake.
The idea is that nothing would be required when updating newmat.

FabienPean replied:
Some points to keep in mind regarding newmat:
* It has not been updated for 9 years or so now.
* It does not offer CMakeLists officially.
* Configuration is meant to be done inside newmat/include.h

Thus the CMakeLists available has been created by Sofa developers, and newmat internals have been modified by Sofa developers too. So if there is an update, the include.h will have to be updated, and the CMakeLists file will remain the same.

Therefore is it still required to have it above in the build tree ? (It would then be located in .\sofa\src\SofaKernel\extlibs\CMakeLists.txt )

matthieu-nesme replied:
ok got it, thanks for your remarks.
Let's merge it then!

matthieu-nesme commented on modules/SofaGeneralEngine/DilateEngine.cpp:
```diff
@@ -38,18 +38,18 @@ SOFA_DECL_CLASS(DilateEngine)
+.add< DilateEngine<Vec3fTypes>>(true)
+#endif //SOFA_FLOAT
 #ifdef SOFA_WITH_DOUBLE
 .add< DilateEngine<Vec3dTypes>>(true) // default template
```
warning: two default templates when compiling with both float and double

damienmarchal commented:
Rebuild from a recent master... [ci-build]

FabienPean commented:
My bad it seems I was quite asleep when I tried to fix the DilateEngine.cpp

damienmarchal commented:
So it seems there is some troubles here in compiling this PR. Anyone can check ?

FabienPean commented:
The problem comes when using 'both' it seems. For one part of the build failure, TriangleOctree / TriangleOctreeRoot must be instantiated with both precision instead of currently using the default type SReal only.

fredroy commented:
I built it on VS2015, and got the same behavior as on the pictures.
It appears to be a bug with SparseGridRamification (in caduceus.scn)
If you replace it with SparseGridTopology, everything appears fine (and behaves fine)

FabienPean committed:
[SofaGeneralEngine_test] Fix IndicesFromValues_test to work with double or float

FabienPean committed:
[SofaGeneralEngine] Fix DilateEngine in case "float" is selected

FabienPean committed:
[SofaGeneralEngine] Fix missing include guard in DilateEngine.cpp

FabienPean committed:
[SofaGeneralEngine_test] Deactivate DilateEngine_test when only float precision selected

damienmarchal commented:
Many thanks @FabienPean for this PR as it seems quite complex to fix everything. 
[ci-build]

FabienPean commented:
Ok so, hopefully the end of this PR. I reverted some commit and double/both should be back to normal. In the case of float only the DilateEngine is not available and its test deactivated. So basically it should compile fine, but some components would be limited.
I quickly tried to fix it but I encountered some difficulties. I would prefer open a new issue regarding float compliance of remaining components.

To sum up, it should be back to the original goal: fix compilation, not necessarily the components with type limitation

damienmarchal commented:
The CI disagree with you :)

FabienPean commented:
I am on Windows and did not try on Linux, so my hopes still hold as long as that yellow dot does not become a cross for the windows VS build :D

The error on the linux build seem trivial fortunately. However it stopped at the first error... Would it be possible for the build system not to stop at this first error and try compiling the remaining ? Since I cannot test on every platform/os that would be valuable to have all potential error at once.

matthieu-nesme commented:
Regarding the visualization problems when compiling with float only: keep in mind that double precision can be necessary in specific situations, so be careful when replacing some hard-coded `double` with `SReal`. Some `double` must remain `double` and I guess some `SReal` should become `double` also in the actual code.
On platforms where double precision does not exist, tricky numerical solutions have to be employed.

FabienPean committed:
[SofaBaseMechanics] Fix DiagonalMass for unix compilation

guparan commented:
I rebased your branch to clean it a bit (removing all merges and reverts).
Let's see how it builds...
https://www.sofa-framework.org/dash/index.php?branch=pr%2Ffix%2Fissue%232

guparan commented:
[ci-build]

guparan committed:
Merge remote-tracking branch 'origin/master' into #262

# Conflicts:
#	CHANGELOG.md

guparan commented:
Are we good to go with this PR? 
@matthieu-nesme, do you see a specific line where a `SReal` should remain a `double`?

matthieu-nesme commented:
There is no easy answer to that, numerical limits are really complex.
But obviously, some double are mandatory somewhere in the components that map the snake model ^^

damienmarchal commented:
[ci-build]
